Add support for LT8722

rccalam opened the pull request:
## PR Description
- This adds driver support and device tree bindings for LT8722 full bridge DC/DC converter.
- Datasheet: [LT8722](https://www.analog.com/media/en/technical-documentation/data-sheets/lt8722.pdf)
- Tested on RPI4 using the following demo board: [DC3145A](https://www.analog.com/en/resources/evaluation-hardware-and-software/evaluation-boards-kits/dc3145a.html)

## PR Type
- [ ] Bug fix (a change that fixes an issue)
- [x] New feature (a change that adds new functionality)
- [ ] Breaking change (a change that affects other repos or cause CIs to fail)

## PR Checklist
- [x] I have conducted a self-review of my own code changes
- [x] I have tested the changes on the relevant hardware
- [ ] I have updated the documentation outside this repo accordingly (if there is the case)

rccalam committed:
dt-bindings: regulator: Add adi,lt8744-regulator.yaml

Add documentation for devicetree bindings for LT8722.

Signed-off-by: Ramon Cristopher M. Calam <ramoncristopher.calam@analog.com>

rccalam committed:
regulator: lt8722: Add driver for LT8722

Add ADI LT8722 full bridge DC/DC converter driver support.

Signed-off-by: Ramon Cristopher M. Calam <ramoncristopher.calam@analog.com>

rccalam committed:
kconfig.adi: Add LT8722

imply REGULATOR_LT8722.

Signed-off-by: Ramon Cristopher M. Calam <ramoncristopher.calam@analog.com>

nunojsa commented on drivers/regulator/lt8722-regulator.c:
```diff
@@ -0,0 +1,641 @@
+				       &chip->vcc_vreg_mV);
+	if (ret == 0) {
+		vcc_vreg = find_closest(chip->vcc_vreg_mV, lt8722_vcc_vreg,
+					ARRAY_SIZE(lt8722_vcc_vreg));
```
Don't hide possible issues with `find_closest()`. Properly validate the value received from DT and error out in case it's invalid.

nunojsa commented on drivers/regulator/lt8722-regulator.c:
```diff
@@ -0,0 +1,641 @@
+		power_limit = 0x01;
+		for (i = 0; i < ARRAY_SIZE(lt8722_power_limit); i++) {
+			if (lt8722_power_limit[i] == chip->power_limit_mW)
+				power_limit = i;
```
no break? And if not find a match?

nunojsa commented on drivers/regulator/lt8722-regulator.c:
```diff
@@ -0,0 +1,641 @@
+					    power_limit);
+		if (ret < 0)
+			return ret;
+	}
```
I would add a dedicated function to parse DT. You have a significant number of properties so I would add a  lt8722_parse_fw() kind of helper

nunojsa commented on drivers/regulator/lt8722-regulator.c:
```diff
@@ -0,0 +1,641 @@
+	if (ret < 0)
+		return ret;
+
+	udelay(160);
```
does it have to be atomic? Hint: use fsleep()

nunojsa commented on drivers/regulator/lt8722-regulator.c:
```diff
@@ -0,0 +1,641 @@
+	}
+
+	// Enable the PWM switching behavior
+	gpiod_set_value(chip->swen_gpio, 1);
```
gpiod_set_value_cansleep()

nunojsa commented on drivers/regulator/lt8722-regulator.c:
```diff
@@ -0,0 +1,641 @@
+		if (ret < 0)
+			return ret;
+
+		mdelay(1);
```
fsleep()

nunojsa commented on drivers/regulator/lt8722-regulator.c:
```diff
@@ -0,0 +1,641 @@
+	mdelay(1);
+
+	// Ramp the output voltage control DAC from 0xFF000000 to 0x00000000
+	for (i = 0; i < 5; i++) {
```
don't hardcode 5.

nunojsa commented on drivers/regulator/lt8722-regulator.c:
```diff
@@ -0,0 +1,641 @@
+	int ret, i;
+
+	if (!chip->spi)
+		return -ENODEV;
```
this is an internal API. The above condition is not really possible unless you call this before the assignment in probe() which is not the case. Hence, drop the check. Same for all other places.

nunojsa commented on drivers/regulator/lt8722-regulator.c:
```diff
@@ -0,0 +1,641 @@
+		return ret;
+
+	return FIELD_GET(LT8722_EN_REQ_MASK, reg_val)
+	       && gpiod_get_value(chip->en_gpio);
```
gpiod_get_value_cansleep()

nunojsa commented on drivers/regulator/lt8722-regulator.c:
```diff
@@ -0,0 +1,641 @@
+		return ret;
+
+	reg_val &= ~mask;
+	reg_val |= (val << __ffs(mask)) & mask;
```
instead of open coding FIELD_PREP(). Can we make sure `val` is already properly set (in the same way we would pass it to regmap_update_bits()) and then have a one liner:

`reg_val = (reg_val & ~mask) | (reg_val & mask);`

nunojsa commented on drivers/regulator/lt8722-regulator.c:
```diff
@@ -0,0 +1,641 @@
+		return -ENODEV;
+
+	u8 rx_buf[LT8722_RW_CMD_SIZE] = {0};
+	u8 tx_buf[LT8722_RW_CMD_SIZE] = {0};
```
Not typical kernel style. All declarations should go before any code.

nunojsa commented on drivers/regulator/lt8722-regulator.c:
```diff
@@ -0,0 +1,641 @@
+		return -ENODEV;
+
+	u8 rx_buf[LT8722_RW_CMD_SIZE] = {0};
+	u8 tx_buf[LT8722_RW_CMD_SIZE] = {0};
```
ditto

nunojsa commented on drivers/regulator/lt8722-regulator.c:
```diff
@@ -0,0 +1,641 @@
+
+static int _lt8722_crc(const u8 *pdata, size_t nbytes)
+{
+	return crc8(lt8722_crc8_table, pdata, nbytes, LT8722_CRC_INIT);
```
Don't think this helper is that helpful

nunojsa commented on drivers/regulator/lt8722-regulator.c:
```diff
@@ -0,0 +1,641 @@
+	tmp = (s64)uV * BIT(LT8722_DAC_BITS);
+	tmp = div_s64(tmp, LT8722_DAC_VREF * 16);
+
+	return clamp_t(s32, tmp, LT8722_MIN_DAC_CODE, LT8722_MAX_DAC_CODE);
```
Hmm when upstreaming make sure you actually need this. All your values should be signed so I think `clamp()` should not complain?

nunojsa reviewed (commented):
here it goes my first round

rccalam commented on drivers/regulator/lt8722-regulator.c:
```diff
@@ -0,0 +1,641 @@
+	ret = device_property_read_u32(&spi->dev, "adi,uv-clamp-microvolt",
+				       &chip->uv_clamp_uV);
+	if (ret == 0) {
+		uv_clamp = FIELD_GET(LT8722_UV_OV_MASK, _lt8722_uV_to_dac(-chip->uv_clamp_uV));
```
We can get the value to be written to the UV_CLAMP register from the 20-23 bits of the DAC code. This is the reason why I used the FIELD_GET to get the UV clamp data.

I will replace this with a look-up table to map the value from the device tree.

rccalam commented:
V2:
	- DT Bindings
		○ Renamed `en-gpios` to `enable-gpios` and `swen-gpios` `switch-enable-gpios` for clarity
		○ Improved descriptions for `enable-gpios` and `switch-enable-gpios to reflect their specific roles
		○ Ensured `description` is the first attribute in property definitions
		○ Refactored `description` remove default value mentions, replacing them with actual default definitions in the schema
		○ Updated `adi,switch-frequency` and `adi,duty-cycle-range` to use descriptive string values for clarity
	- Kconfig
		○ Updated Kconfig to reflect that LT8722 depends on both SPI and OF
	- Driver
		○ Sorted header files alphabetically
		○ Adopted correct kernel style single line comment
		○ Removed extra lines
		○ Removed redundant GPIO direction setting
		○ Replaced ret == 0 with !ret
		○ Moved device tree properties retrieval into a separate helper function.
		○ Used fsleep()
		○ Used gpiod_set_value_cansleep()
		○ Introduced LT8722_RAMP_STEP constant to replace hardcoded value
		○ Removed unnecessary NULL check for SPI device
		○ Ensured all declarations are positioned before any code
		○ Deleted _lt8722_crc helper function
		○ Prepared val before passing value to lt8722_reg_write_mask.
		○ Explicitly cast LT8722_MIN/MAX_DAC_CODE as signed and used clamp() function

rccalam commented:
V2:

**DT Bindings**
- Renamed `en-gpios` to `enable-gpios` and `swen-gpios` `switch-enable-gpios` for clarity
- Improved descriptions for `enable-gpios` and `switch-enable-gpios` to reflect their specific roles
- Ensured `description` is the first attribute in property definitions
- Refactored `description` remove default value mentions, replacing them with actual default definitions in the schema
- Updated `adi,switch-frequency` and `adi,duty-cycle-range` to use descriptive string values for clarity

**Kconfig**
- Updated Kconfig to reflect that LT8722 depends on both SPI and OF

**Driver**
- Sorted header files alphabetically
- Adopted correct kernel style single line comment
- Removed extra lines
- Removed redundant GPIO direction setting
- Replaced ret == 0 with !ret
- Moved device tree properties retrieval into a separate helper function.
- Used `fsleep()`
- Used `gpiod_set_value_cansleep()`
- Introduced `LT8722_RAMP_STEP` constant to replace hardcoded value
- Removed unnecessary NULL check for SPI device
- Ensured all declarations are positioned before any code
- Deleted `_lt8722_crc helper` function
- Prepared val before passing value to `lt8722_reg_write_mask`
- Explicitly cast `LT8722_MIN/MAX_DAC_CODE` as signed and used clamp() function
- Removed SPI setup

rccalam commented on drivers/regulator/lt8722-regulator.c:
```diff
@@ -0,0 +1,641 @@
+
+	ret = gpiod_direction_output(chip->swen_gpio, 0);
+	if (ret < 0)
+		return ret;
```
You're this is redundant. Removed this in V2.

rccalam commented:
V2:

**DT Bindings**
- Renamed `en-gpios` to `enable-gpios` and `swen-gpios` `switch-enable-gpios` for clarity
- Improved descriptions for `enable-gpios` and `switch-enable-gpios` to reflect their specific roles
- Ensured `description` is the first attribute in property definitions
- Refactored `description` remove default value mentions, replacing them with actual default definitions in the schema
- Updated `adi,switch-frequency` and `adi,duty-cycle-range` to use descriptive string values for clarity

**Kconfig**
- Updated Kconfig to reflect that LT8722 depends on both SPI and OF

**Driver**
- Sorted header files alphabetically
- Adopted correct kernel style single line comment
- Removed extra lines
- Removed redundant GPIO direction setting
- Replaced `ret == 0` with `!ret`
- Moved device tree properties retrieval into a separate helper function.
- Used `fsleep()`
- Used `gpiod_set_value_cansleep()`
- Introduced `LT8722_RAMP_STEP` constant to replace hardcoded value
- Removed unnecessary `NULL` check for SPI device
- Ensured all declarations are positioned before any code
- Deleted `_lt8722_crc helper` function
- Prepared val before passing value to `lt8722_reg_write_mask`
- Explicitly cast `LT8722_MIN/MAX_DAC_CODE` as signed and used `clamp()` function
- Removed SPI setup

rccalam commented:
V2:

**DT Bindings**
- Renamed `en-gpios` to `enable-gpios` and `swen-gpios` `switch-enable-gpios` for clarity
- Improved descriptions for `enable-gpios` and `switch-enable-gpios` to reflect their specific roles
- Ensured `description` is the first attribute in property definitions
- Refactored `description` remove default value mentions, replacing them with actual default definitions in the schema
- Updated `adi,switch-frequency` and `adi,duty-cycle-range` to use descriptive string values for clarity

**Kconfig**
- Updated Kconfig to reflect that LT8722 depends on both SPI and OF

**Driver**
- Sorted header files alphabetically
- Adopted correct kernel style single line comment
- Removed extra lines
- Removed redundant GPIO direction setting
- Replaced `ret == 0` with `!ret`
- Moved device tree properties retrieval into a separate helper function.
- Used `fsleep()`
- Used `gpiod_set_value_cansleep()`
- Introduced `LT8722_RAMP_STEP` constant to replace hardcoded value
- Removed unnecessary `NULL` check for SPI device
- Ensured all declarations are positioned before any code
- Deleted `_lt8722_crc helper` function
- Prepared val before passing value to `lt8722_reg_write_mask`
- Removed use of clamp()
- Removed SPI setup

nunojsa commented on Documentation/devicetree/bindings/regulator/adi,lt8722-regulator.yaml:
```diff
@@ -0,0 +1,146 @@
+    maxItems: 1
+
+  enable-gpios:
+    description: |
```
drop the `|`. This does not need to be formatted text. Same for other properties

nunojsa commented on drivers/regulator/lt8722-regulator.c:
```diff
@@ -0,0 +1,716 @@
+	if (IS_ERR(chip->swen_gpio))
+		return PTR_ERR(chip->swen_gpio);
+
+	/* Reset device */
```
nit: redundant comment

nunojsa commented on drivers/regulator/lt8722-regulator.c:
```diff
@@ -0,0 +1,716 @@
+	en_req = FIELD_GET(LT8722_EN_REQ_MASK, reg_val);
+	en_pin = gpiod_get_value(chip->en_gpio);
+
+	return en_req || en_pin;
```
I got the idea this should be `en_req && en_pin`? Maybe I got it wrong

rccalam replied:
You are right, it's supposed to be en_req && en_pin.

nunojsa commented on drivers/regulator/lt8722-regulator.c:
```diff
@@ -0,0 +1,716 @@
+	u8 tx_buf[LT8722_RW_CMD_SIZE] = {0};
+
+	if (!spi)
+		return -ENODEV;
```
The above check should not be needed...

nunojsa commented on drivers/regulator/lt8722-regulator.c:
```diff
@@ -0,0 +1,716 @@
+	u8 tx_buf[LT8722_RW_CMD_SIZE] = {0};
+
+	if (!spi)
+		return -ENODEV;
```
ditto

nunojsa commented on drivers/regulator/lt8722-regulator.c:
```diff
@@ -0,0 +1,716 @@
+	tmp = (s64)uV * BIT(LT8722_DAC_BITS);
+	tmp = div_s64(tmp, LT8722_DAC_VREF * 16);
+
+	return (s32)tmp;
```
There's already an implicit cast. Not need to do it explicitly

nunojsa commented on drivers/regulator/lt8722-regulator.c:
```diff
@@ -0,0 +1,716 @@
+	tmp = (s64)dac_val * LT8722_DAC_VREF;
+	tmp = 16 * div_s64(tmp, BIT(LT8722_DAC_BITS));
+
+	return (s32)tmp;
```
ditto

nunojsa commented on drivers/regulator/lt8722-regulator.c:
```diff
@@ -0,0 +1,716 @@
+	ret = device_property_read_u32(&chip->spi->dev, "adi,uv-clamp-microvolt",
+				       &chip->uv_clamp_uV);
+	if (!ret) {
+		/* Interpret the UV clamp value as a negative number */
```
Not getting this... Why do we need to have it? Note that you can give negative values through DT...

nunojsa commented on drivers/regulator/lt8722-regulator.c:
```diff
@@ -0,0 +1,716 @@
+			return -EINVAL;
+
+		uv_clamp = find_closest(chip->uv_clamp_uV, lt8722_uv_clamp,
+					ARRAY_SIZE(lt8722_uv_clamp));
```
I would not do it like this. Just have an enum in DT with all the possibilities and the n do a lookup here. And error out if a not supported value was given.  Same for the other properties where this applies.

nunojsa commented on drivers/regulator/lt8722-regulator.c:
```diff
@@ -0,0 +1,716 @@
+				break;
+
+		if (i == ARRAY_SIZE(lt8722_power_limit))
+			return -EINVAL;
```
I guess you could come up with a generic helper for this lookups in the static arrays...

nunojsa commented on drivers/regulator/lt8722-regulator.c:
```diff
@@ -0,0 +1,716 @@
+					    FIELD_PREP(LT8722_SYS_DC_MASK, i));
+		if (ret < 0)
+			return ret;
+	}
```
The strings properties can be done much better. Hint:

[device_property_match_property_string()](https://elixir.bootlin.com/linux/v6.12-rc5/C/ident/device_property_match_property_string)

nunojsa reviewed (commented):
Looks better. take my inputs and send it upstream... I think it's already good enough to be sent. Of course you'll get more feedback from the maintainers.

nunojsa commented:
@rccalam, is this sent? What is the status on this one?

nunojsa commented:
@rccalam Do you plan on having this one done or should I close this?
